skbio/metadata/_interval.py

```diff
@@ -0,0 +1,334 @@
+
+        """
+        # Add an interval to the tree. Note that the add functionality is
+        # built within the Interval constructor.
+        Interval(interval_metadata=self,
+                 intervals=intervals,
+                 boundaries=boundaries,
+                 metadata=metadata)
+
+    def _rebuild_tree(self, intervals):
+        """ Rebuilds the IntervalTree when the tree is stale."""
+        self._intervals = IntervalTree()
+        for f in intervals:
+            for inv in f.intervals:
+                start, end = inv
+                self._intervals.add(start, end, f)
+
+    def _query_interval(self, interval):
+        """ Fetches Interval objects based on query interval"""
+        start, end = _polish_interval(interval)
+        invs = self._intervals.find(start, end)
+        return invs
+
+    def _query_attribute(self, intervals, metadata):
+        """ Fetches Interval objects based on query attributes"""
+        if metadata is None:
+            return []
+
+        queries = []
+        for inv in intervals:
+            for (key, value) in metadata.items():
+                if inv[key] != value:
+                    continue
+                queries.append(inv)
+        return queries
+
+    def query(self, intervals=None, boundaries=None, metadata=None):
+        """ Looks up Interval objects with the intervals, boundaries and keywords.
+
+        Parameters
+        ----------
+        intervals : iterable of tuple of ints
+            A list of intervals associated with the Interval object.
+        boundaries : iterable of tuple of bool
+            A list of boundaries associated with the Interval object.
+        metadata : dict
+            A dictionary of key word attributes associated with the
+            Interval object.
+
+        Returns
+        -------
+        list, Interval
+            A list of Interval objects satisfying the search criteria.
+
+        Examples
+        --------
+        >>> from skbio.metadata import IntervalMetadata
+        >>> interval_metadata = IntervalMetadata()
+        >>> interval_metadata.add(intervals=[(0, 2), (4, 7)],
+        ...                       boundaries=None, metadata={'name': 'sagA'})
+        >>> interval_metadata.add(intervals=[(40, 70)],
+        ...                       boundaries=None, metadata={'name': 'sagB'})
+        >>> interval_metadata.query(intervals=[(1, 2)])
+        [Interval(intervals=[(0, 2), (4, 7)], metadata={'name': 'sagA'})]
+
+        Note
+        ----
+        There are two types of queries to perform
+        1. Query by interval.
+        2. Query by key/val pair (i.e. gene=sagA).
+
+        """
+        if self._is_stale_tree:
+            self._rebuild_tree(self._metadata)
+            self._is_stale_tree = False
+
+        invs = set()
+
+        # Find queries by interval
+        if intervals is not None:
+            for value in intervals:
+                invs.update(self._query_interval(value))
```

Comment:
@mortonjt, is this the motivator for doing a hashable above? would it be sufficient to assert non-redundancy in `intervals` instead of relying on a `set`?